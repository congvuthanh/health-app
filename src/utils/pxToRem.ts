/**
 * Converts pixel values to rem units based on 10px base (html font-size: 62.5%)
 * @param px - Pixel value to convert
 * @returns rem value as string
 * @example
 * pxToRem(24) // returns '2.4rem'
 * pxToRem(16) // returns '1.6rem'
 */
export const pxToRem = (px: number): string => {
  return `${px / 10}rem`;
};

/**
 * Converts multiple pixel values to rem units (useful for shorthand properties)
 * @param values - Array of pixel values
 * @returns Space-separated rem values as string
 * @example
 * pxToRem(10, 20, 30) // returns '1rem 2rem 3rem'
 */
export const pxToRemMultiple = (...values: number[]): string => {
  return values.map((value) => `${value / 10}rem`).join(' ');
};
