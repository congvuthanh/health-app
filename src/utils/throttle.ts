/* eslint-disable @typescript-eslint/no-this-alias */
/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 * Creates a throttled function that only invokes the provided function at most once
 * per every `wait` milliseconds.
 *
 * @param func - The function to throttle
 * @param wait - The number of milliseconds to throttle invocations to
 * @returns A throttled version of the function
 *
 * @example
 * const handleScroll = throttle(() => {
 *   console.log('Scrolled!');
 * }, 200);
 *
 * window.addEventListener('scroll', handleScroll);
 */
export function throttle<T extends (...args: any[]) => any>(
  func: T,
  wait: number
): (...args: Parameters<T>) => void {
  let timeout: ReturnType<typeof setTimeout> | null = null;
  let lastRan: number | null = null;

  return function (this: any, ...args: Parameters<T>) {
    const context = this;
    const now = Date.now();

    if (!lastRan) {
      // First call - execute immediately
      func.apply(context, args);
      lastRan = now;
    } else {
      const timeSinceLastRan = now - lastRan;

      if (timeSinceLastRan >= wait) {
        // Enough time has passed - execute immediately
        func.apply(context, args);
        lastRan = now;
      } else {
        // Clear any existing timeout
        if (timeout) {
          clearTimeout(timeout);
        }

        // Set a new timeout to execute after the remaining wait period
        timeout = setTimeout(() => {
          func.apply(context, args);
          lastRan = Date.now();
          timeout = null;
        }, wait - timeSinceLastRan);
      }
    }
  };
}
