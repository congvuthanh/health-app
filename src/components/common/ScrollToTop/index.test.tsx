import { act, renderWithProviders, screen, waitFor } from 'utils/test';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { ScrollToTop } from './index';

describe('ScrollToTop', () => {
  const renderComponent = () => {
    return renderWithProviders(<ScrollToTop />);
  };

  beforeEach(() => {
    window.scrollTo = vi.fn();
  });

  describe('Rendering', () => {
    it('renders scroll to top button', () => {
      renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      expect(button).toBeInTheDocument();
      expect(button).toHaveAttribute('aria-label', 'ページのトップへ戻る');
    });

    it('renders as a button element', () => {
      renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      expect(button.tagName).toBe('BUTTON');
    });

    it('has type button attribute', () => {
      renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      expect(button).toHaveAttribute('type', 'button');
    });
  });

  describe('Functionality', () => {
    it('scrolls to top when clicked', async () => {
      const { user } = renderComponent();

      // Make button visible first by scrolling down
      act(() => {
        Object.defineProperty(window, 'scrollY', {
          writable: true,
          configurable: true,
          value: 300,
        });
        window.dispatchEvent(new Event('scroll'));
      });

      const button = screen.getByRole('button', { hidden: true });
      await user.click(button);

      expect(window.scrollTo).toHaveBeenCalledWith({
        top: 0,
        behavior: 'smooth',
      });
    });

    it('calls scrollTo only once per click', async () => {
      const { user } = renderComponent();

      // Make button visible first by scrolling down
      act(() => {
        Object.defineProperty(window, 'scrollY', {
          writable: true,
          configurable: true,
          value: 300,
        });
        window.dispatchEvent(new Event('scroll'));
      });

      const button = screen.getByRole('button', { hidden: true });
      await user.click(button);

      expect(window.scrollTo).toHaveBeenCalledTimes(1);
    });

    it('scrolls to top on multiple clicks', async () => {
      const { user } = renderComponent();

      // Make button visible first by scrolling down
      act(() => {
        Object.defineProperty(window, 'scrollY', {
          writable: true,
          configurable: true,
          value: 300,
        });
        window.dispatchEvent(new Event('scroll'));
      });

      const button = screen.getByRole('button', { hidden: true });

      await user.click(button);
      await user.click(button);
      await user.click(button);

      expect(window.scrollTo).toHaveBeenCalledTimes(3);
      expect(window.scrollTo).toHaveBeenLastCalledWith({
        top: 0,
        behavior: 'smooth',
      });
    });
  });

  describe('Accessibility', () => {
    it('has proper aria-label', () => {
      renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      expect(button).toHaveAttribute('aria-label', 'ページのトップへ戻る');
    });

    it('is keyboard accessible', () => {
      renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      button.focus();
      expect(button).toHaveFocus();
    });

    it('can be activated with Enter key', async () => {
      const { user } = renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      button.focus();

      await user.keyboard('{Enter}');

      expect(window.scrollTo).toHaveBeenCalled();
    });

    it('can be activated with Space key', async () => {
      const { user } = renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      button.focus();

      await user.keyboard(' ');

      expect(window.scrollTo).toHaveBeenCalled();
    });
  });

  describe('Visibility based on scroll position', () => {
    beforeEach(() => {
      // Reset scroll position
      Object.defineProperty(window, 'scrollY', {
        writable: true,
        configurable: true,
        value: 0,
      });
    });

    it('is hidden when scroll position is below threshold (200px)', () => {
      renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      const styles = window.getComputedStyle(button);

      expect(styles.opacity).toBe('0');
      expect(styles.visibility).toBe('hidden');
    });

    it('shows button when scrolled past threshold', async () => {
      renderComponent();

      // Simulate scrolling past threshold
      act(() => {
        Object.defineProperty(window, 'scrollY', {
          writable: true,
          configurable: true,
          value: 250,
        });
        window.dispatchEvent(new Event('scroll'));
      });

      // Wait for throttled function to execute
      await waitFor(() => {
        const button = screen.getByRole('button', { hidden: true });
        const styles = window.getComputedStyle(button);
        expect(styles.opacity).toBe('1');
      });

      const button = screen.getByRole('button', { hidden: true });
      const styles = window.getComputedStyle(button);
      expect(styles.visibility).toBe('visible');
    });

    it('hides button when scrolled back to top', async () => {
      renderComponent();

      // Scroll down first
      act(() => {
        Object.defineProperty(window, 'scrollY', {
          writable: true,
          configurable: true,
          value: 300,
        });
        window.dispatchEvent(new Event('scroll'));
      });

      // Wait for button to become visible
      await waitFor(() => {
        const button = screen.getByRole('button', { hidden: true });
        const styles = window.getComputedStyle(button);
        expect(styles.opacity).toBe('1');
      });

      // Then scroll back to top
      act(() => {
        Object.defineProperty(window, 'scrollY', {
          writable: true,
          configurable: true,
          value: 100,
        });
        window.dispatchEvent(new Event('scroll'));
      });

      // Wait for button to become hidden
      await waitFor(() => {
        const button = screen.getByRole('button', { hidden: true });
        const styles = window.getComputedStyle(button);
        expect(styles.opacity).toBe('0');
      });

      const button = screen.getByRole('button', { hidden: true });
      const styles = window.getComputedStyle(button);
      expect(styles.visibility).toBe('hidden');
    });

    it('updates visibility at exactly 200px threshold', async () => {
      renderComponent();

      // At threshold
      act(() => {
        Object.defineProperty(window, 'scrollY', {
          writable: true,
          configurable: true,
          value: 200,
        });
        window.dispatchEvent(new Event('scroll'));
      });

      let button = screen.getByRole('button', { hidden: true });
      let styles = window.getComputedStyle(button);
      expect(styles.opacity).toBe('0');

      // Just past threshold
      act(() => {
        Object.defineProperty(window, 'scrollY', {
          writable: true,
          configurable: true,
          value: 201,
        });
        window.dispatchEvent(new Event('scroll'));
      });

      // Wait for throttled function to execute
      await waitFor(() => {
        button = screen.getByRole('button', { hidden: true });
        styles = window.getComputedStyle(button);
        expect(styles.opacity).toBe('1');
      });
    });

    it('disables pointer events when hidden', () => {
      renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      const styles = window.getComputedStyle(button);

      expect(styles.pointerEvents).toBe('none');
    });

    it('enables pointer events when visible', async () => {
      renderComponent();

      act(() => {
        Object.defineProperty(window, 'scrollY', {
          writable: true,
          configurable: true,
          value: 300,
        });
        window.dispatchEvent(new Event('scroll'));
      });

      // Wait for throttled function to execute
      await waitFor(() => {
        const button = screen.getByRole('button', { hidden: true });
        const styles = window.getComputedStyle(button);
        expect(styles.pointerEvents).toBe('auto');
      });
    });
  });

  describe('Styling', () => {
    it('has fixed position', () => {
      renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      const styles = window.getComputedStyle(button);
      expect(styles.position).toBe('fixed');
    });

    it('has circular shape (border-radius)', () => {
      renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      const styles = window.getComputedStyle(button);
      expect(styles.borderRadius).toBe('100%');
    });

    it('has pointer cursor', () => {
      renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      const styles = window.getComputedStyle(button);
      expect(styles.cursor).toBe('pointer');
    });

    it('has appropriate z-index', () => {
      renderComponent();
      const button = screen.getByRole('button', { hidden: true });
      const styles = window.getComputedStyle(button);
      expect(styles.zIndex).toBe('9000');
    });
  });

  describe('Snapshot', () => {
    it('matches snapshot', () => {
      const { container } = renderComponent();
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
