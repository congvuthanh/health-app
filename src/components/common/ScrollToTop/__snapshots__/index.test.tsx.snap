// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ScrollToTop > Snapshot > matches snapshot 1`] = `
<button
  aria-label="ページのトップへ戻る"
  class="sc-bRKDuR fJyhPG"
  type="button"
>
  <img
    alt="default"
    class="sc-hvigdm cGmgfQ"
    src="data:image/svg+xml,%3csvg%20width='48'%20height='48'%20viewBox='0%200%2048%2048'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M24%200.5C36.9787%200.5%2047.5%2011.0213%2047.5%2024C47.5%2036.9787%2036.9787%2047.5%2024%2047.5C11.0213%2047.5%200.5%2036.9787%200.5%2024C0.5%2011.0213%2011.0213%200.5%2024%200.5Z'%20stroke='%23414141'%20stroke-opacity='0.8'/%3e%3cpath%20d='M30.5853%2028.042L24.0003%2021.6579L17.4153%2028.042L16.5391%2027.1925L24.0003%2019.959L31.4615%2027.1925L30.5853%2028.042Z'%20fill='%23414141'%20fill-opacity='0.8'/%3e%3c/svg%3e"
  />
  <img
    alt="hover"
    class="sc-fhHczv iealsO"
    src="data:image/svg+xml,%3csvg%20width='48'%20height='48'%20viewBox='0%200%2048%2048'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M24%200C37.2548%200%2048%2010.7452%2048%2024C48%2037.2548%2037.2548%2048%2024%2048C10.7452%2048%200%2037.2548%200%2024C0%2010.7452%2010.7452%200%2024%200Z'%20fill='%23FF963C'/%3e%3cpath%20d='M30.5853%2028.042L24.0003%2021.6579L17.4153%2028.042L16.5391%2027.1925L24.0003%2019.959L31.4615%2027.1925L30.5853%2028.042Z'%20fill='white'%20stroke='white'/%3e%3c/svg%3e"
  />
</button>
`;
